### DISCUSSION TITLE
std::vector and oder containers as userdata

### DISCUSSION THREAD

**Discussion by FilippoLeon** [kunitoki/LuaBridge3#176] [Q&A]
So I'd like to expose automatically std::vector<T> and similar to luabridge, but not as lua table (like including Vector.h) (which incurs in a lot of copies and performance impacts for big vectors), but just a userdata with accessors. 

Is there a way to do this generically without having to register every single instance manually?

Also, ideally I would also support the same syntax lua tables, I think it is easy to overload # and index metamethods, however, if I want to implement ipairs support I need to have multiple return values from luabridge. 

Any idea about how best to do it?

**Comment by kunitoki** [comment]
I can try to come up with some code later how to do it easily.

**Comment by FilippoLeon** [comment]
I can give it a try myself, could be fun. I just need a starting point.

**Comment by kunitoki** [comment]
An initial starting point:

```cpp
#include <vector>

template <class T>
void registerVector(luabridge::Namespace luaNamespace, const char* className)
{
    using U = std::vector<T>;

    luaNamespace
        .beginClass<U>(className)
            .addConstructor([](void* ptr) { return new(ptr) U(); })
            .addFunction("push_back", luabridge::nonConstOverload<const float&>(&U::push_back))
            .addFunction("at", luabridge::constOverload<std::size_t>(&U::at))
            .addFunction("size", &U::size)
        .endClass()
        .endNamespace();
}

TEST_F(VectorTests, RegisterCustomTest)
{
    auto localNamespace = luabridge::getGlobalNamespace(L).beginNamespace("test");

    registerVector<float>(std::move(localNamespace), "FloatVector");

    runLua("a = test.FloatVector(); a:push_back(1.0); result = a:size()");
    EXPECT_EQ(1, result<int>());
}
```

**Reply by FilippoLeon** [reply]
Actually I was thinking about a way to have it registered automatically. Say I have a class with a std::vector<int> member, and I expose this. If I include Vector.h then everything is cool, but I get a lua table that uses the stack. If I leave it alone, I get an unregistered userdata that I have to find out myself and register. 

What I'd like to have is a way to combine both worlds, and have a usertype that is generically registered once, if that makes sense.

**Reply by kunitoki** [reply]
It's hard for luabridge to understand when a `std::vector<T>` should be used as a lua table or as userdata.

If you registered a `FloatVector` type in luabridge it should be automatically found when you expose a `std::vector<float>` to lua, but you shouldn't include Vector.h as that will make it think the `std::vector<float>` is a convertible to lua table.

**Reply by totalgee** [reply]
This (`registerVector<type>()`) makes sense, it's similar to what I'm used to from ChaiScript, where there are helpers to "bootstrap" various std library containers (actually, this is initially what I expected including `Vector.h` would provide). In ChaiScript, you need to explicitly register specific container types you want exposed to the scripting language (for each templated type), but it's as easy as a call to a helper, similar to what you propose above:

https://github.com/ChaiScript/ChaiScript/blob/develop/include/chaiscript/dispatchkit/bootstrap_stl.hpp#L425

**Reply by totalgee** [reply]
@kunitoki I tried your `registerVector()` helper, and it seems to work for the basics (adding an element). One minor thing to change is this line:

```c++
            .addFunction("push_back", luabridge::nonConstOverload<const T&>(&U::push_back))
```
(use `<const T&>` as shown as the template parameter, rather than `<const float&>` as you had in your example)

But what more is needed to expose via `registerVector()`, to allow iteration over this vector type in Lua (e.g. using `for k = 1, #vec do ...`)? If I expose `__len` to give the result of `std::vector::size()`, then calling `#vec` in Lua seems to work to get the size)... How can I do something similar for indexing? I don't know enough Lua to be sure what's the best way to do this for userdata... I tried exposing a version of `__index` as shown below, but so far only caused crashes. (-;

```c++
    template <class T>
    void registerVector(luabridge::Namespace luaNamespace, const char* className)
    {
        using U = std::vector<T>;

        luaNamespace
            .beginClass<U>(className)
            .addConstructor([](void* ptr) { return new(ptr) U(); })
            .addFunction("push_back", luabridge::nonConstOverload<T const&>(&U::push_back))
            .addFunction("at", luabridge::constOverload<std::size_t>(&U::at))
            .addFunction("size", &U::size)
            // The following seems to work, so #vec gives the correct size in Lua:
            .addFunction("__len", &U::size)
            // The following is not correct to allow vec[n] indexing; I get a crash on push_back():
            //.addFunction("__index", luabridge::nonConstOverload<std::size_t>(&U::at))
            .endClass()
            .endNamespace();
    }
```

**Comment by kunitoki** [comment]
There are some tests that show how to deal with __index and __newindex metamethods registration https://github.com/kunitoki/LuaBridge3/blob/master/Tests/Source/ClassTests.cpp#L2328

You'll probably have to register a lambda that matches the way lua table indexing work.

As this is a request, i will look at providing these stl containers as userdata registration methods.
